Add M+ Keystone to GameMenu

billriess opened the pull request:
This will display the current M+ Keystone for players who are max level.

<!--
Creating the PR.
- Fill the template, add/remove sections as needed.
-->

Improves #159 

<!-- If needed, link to design -->

# Summary of Changes

1. Checks if you are max level.
2. Shows "M+ Keystone: <dungeonName> (+<dungeonLevel>)" if you have a keystone in your bags.
3. Shows "M+ Keystone: None" if you don't have a keystone.

# Description
Shows your current keystone.

# To test
Hit ESC with a keystone in your bag, no keystone or under max level.

billriess committed:
Add M+ Keystone to GameMenu

This will display the current M+ Keystone for players who are max level.

billriess commented:
I should note that this may be a good option for WunderBar as well or alternatively. It would work well in the XP space since that section is usually blank unless tracking a specific rep.

Toxicom commented:
Thanks for opening a PR! Looks good, although there are a couple things before I merge:
- This needs to be retail only, as I'm fairly certain `C_MythicPlus` will just crash on cata/classic
- The keystone doesn't seem to update dynamically, meaning I have to `/reload` each time my keystone changes
![image](https://github.com/user-attachments/assets/5ecd1d08-333b-4833-b140-1f62d6934594)

billriess commented:
Oops you’re right. I don’t play classic so wasn’t thinking about that. I’ll add logic around checking for that and updating the key on change.

billriess committed:
Update GameMenuButton.lua

Improved compatibility with non-retail clients and listen for events to update the keystone text.

billriess commented:
I've updated the logic to check for retail client as well as update based on several keystone events and bag updates.

billriess committed:
Update GameMenuButton.lua

Fixed overlapping keystone text on change.

Toxicom commented on Modules/Misc/GameMenuButton.lua:
```diff
@@ -108,6 +108,44 @@ function M:GameMenuButton()
         collections.achievs:SetTextColor(1, 1, 1, 1)
+
+        -- Mythic+ Keystone
+        if WOW_PROJECT_ID == WOW_PROJECT_MAINLINE then
```
```suggestion
        if TXUI.IsRetail then
```

Toxicom commented on Modules/Misc/GameMenuButton.lua:
```diff
@@ -108,6 +108,44 @@ function M:GameMenuButton()
+
+        -- Mythic+ Keystone
+        if WOW_PROJECT_ID == WOW_PROJECT_MAINLINE then
+          local keystoneFrame = CreateFrame("Frame")
+          keystoneFrame:RegisterEvent("BAG_UPDATE")
+          keystoneFrame:RegisterEvent("CHALLENGE_MODE_COMPLETED")
+          keystoneFrame:RegisterEvent("CHALLENGE_MODE_RESET")
+          keystoneFrame:RegisterEvent("CHALLENGE_MODE_START")
+          keystoneFrame:RegisterEvent("PLAYER_ENTERING_WORLD")
```
I don't think it's necessary to register all these events for displaying a single string. The whole string display logic can be moved further down below to the `OnShow` hook and simply update the keystone each time menu is opened.

billriess commented:
I can make these changes later. As I said I’m fairly new to lua and wow addons in general. I’m just using the API reference and learning as I go. This makes sense though.

Sorry if I’m causing you more work having to review these. Just trying to contribute.

Toxicom commented:
All good, I could do these changes myself but figured you wanna learn haha

billriess committed:
Update GameMenuButton.lua

Reworked logic to check onshow instead of events.

billriess commented:
I was going to add a `TXUI.IsMaxLevel` helper but wasn't sure how useful that would be for you. Might be handy if more end-game or M+ related things are added.

I've updated the rest of the code lmk what you think.

Toxicom commented:
> I was going to add a `TXUI.IsMaxLevel` helper but wasn't sure how useful that would be for you. Might be handy if more end-game or M+ related things are added.

We do already have a table for that: https://github.com/Toxicom/toxiui/blob/development/Core/Internal.lua#L19-L23
And I see DataBar is using `IsPlayerAtEffectiveMaxLevel` https://github.com/Toxicom/toxiui/blob/development/Modules/WunderBar/SubModules/DataBar.lua#L394

I'll most likely merge this in today, thanks again!
